tx-parser/src/lib.rs

```diff
@@ -83,86 +95,91 @@ def_parser_fn!(Tx, temp_id, Value, String, input, {
+def_parser_fn!(Tx, atom, Value, Value, input, {
+    satisfy_map(|x: Value| if x.is_atom() { Some(x) } else { None })
         .parse_stream(input)
 });
 
-def_parser_fn!(Tx, retract, Value, Entity, input, {
-    satisfy_map(|x: Value| -> Option<Entity> {
-            if let Value::Vector(y) = x {
-                let mut p = (token(Value::NamespacedKeyword(NamespacedKeyword::new("db", "retract"))),
-                             Tx::<&[Value]>::entid_or_lookup_ref_or_temp_id(),
-                             Tx::<&[Value]>::entid(),
-                             // TODO: handle lookup-ref.
-                             any(),
-                             eof())
-                    .map(|(_, e, a, v, _)| {
-                        Entity::AddOrRetract {
-                            op: OpType::Retract,
-                            e: e,
-                            a: a,
-                            v: v,
-                        }
-                    });
-                // TODO: use ok() with a type annotation rather than explicit match.
-                match p.parse_lazy(&y[..]).into() {
-                    Ok((r, _)) => Some(r),
-                    _ => None,
-                }
-            } else {
-                None
-            }
-        })
-        .parse_stream(input)
+fn value_to_nested_vector(val: &Value) -> Option<Vec<AtomOrLookupRefOrVectorOrMapNotation>> {
+    val.as_vector().and_then(|vs| {
+        let mut p = (many(Tx::<&[Value]>::atom_or_lookup_ref_or_vector()),
+                     eof())
+            .map(|(vs, _)| vs);
+        let r: ParseResult<Vec<AtomOrLookupRefOrVectorOrMapNotation>, _> =  p.parse_lazy(&vs[..]).into();
+        r.map(|x| x.0).ok()
+    })
+}
+def_value_satisfy_parser_fn!(Tx, nested_vector, Vec<AtomOrLookupRefOrVectorOrMapNotation>, value_to_nested_vector);
+
+def_parser_fn!(Tx, atom_or_lookup_ref_or_vector, Value, AtomOrLookupRefOrVectorOrMapNotation, input, {
+    Tx::<I>::lookup_ref().map(AtomOrLookupRefOrVectorOrMapNotation::LookupRef)
+        .or(Tx::<I>::nested_vector().map(AtomOrLookupRefOrVectorOrMapNotation::Vector))
+        .or(Tx::<I>::map_notation().map(AtomOrLookupRefOrVectorOrMapNotation::MapNotation))
+        .or(Tx::<I>::atom().map(AtomOrLookupRefOrVectorOrMapNotation::Atom))
+        .parse_lazy(input)
+        .into()
 });
 
+fn value_to_add_or_retract(val: &Value) -> Option<Entity> {
+    val.as_vector().and_then(|vs| {
+        let add = token(Value::NamespacedKeyword(NamespacedKeyword::new("db", "add")))
+            .map(|_| OpType::Add);
+        let retract = token(Value::NamespacedKeyword(NamespacedKeyword::new("db", "retract")))
+            .map(|_| OpType::Retract);
+        let mut p = (add.or(retract),
+                     Tx::<&[Value]>::entid_or_lookup_ref_or_temp_id(),
+                     Tx::<&[Value]>::entid(),
+                     Tx::<&[Value]>::atom_or_lookup_ref_or_vector(),
+                     eof())
+            .map(|(op, e, a, v, _)| {
```

Comment:
Per Marwes, the pattern is (...).skip(eof()), which I'll start using in this file.